File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length,weight,truck_weights):
    time=0
    bridge=deque([0]*bridge_length)
    truck_weights=deque(truck_weights)
        
    current=0
    #다리에서 먼저 빼고난후 들어올 truckweight.popleft()와 더해서 비교
    while len(truck_weights)!=0:
        time+=1
        current-=bridge.popleft()
            
        if current+truck_weights[0]<=weight:
            current+=truck_weights[0]
            bridge.append(truck_weights.popleft())
        else:
            bridge.append(0)
    time+=bridge_length
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

public class Solution {
        public int solution(int bridge_length, int weight, int[] truck_weights) {
            Deque<Integer> bridge = new ArrayDeque<>();
            int time = 0;
            int currentWeight = 0;
            int index = 0;

            for (int i = 0; i < bridge_length; i++) {
                bridge.add(0);
            }

            while (!bridge.isEmpty()) {
                time++;
                currentWeight -= bridge.poll();

                if (index < truck_weights.length) {
                    if (currentWeight + truck_weights[index] <= weight) {
                        bridge.add(truck_weights[index]);
                        currentWeight += truck_weights[index];
                        index++;
                    } else {
                        bridge.add(0);
                    }
                }
            }

            return time;
        }
}
